### DISCUSSION TITLE
How to split data into train, val and test for link prediction in MetaPath2Vec model on Aminer dataset?

### DISCUSSION THREAD

**Discussion by xubingze** [pyg-team/pytorch_geometric#8875] [Q&A] [answered]
Hi, I am trying to do link prediction with the MetaPath2Vec model and I found the example [here](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/hetero/metapath2vec.py).

However, in my dataset I do not have any node labels so I need to split the data into train, val and test, and only use train_data for "train" function and model loader. I am trying to do the link prediction with AMiner dataset but I encountered the following error. Can you help me with the problem? Thank you very much!

```
import os.path as osp
import torch_geometric.transforms as T
import torch

from torch_geometric.datasets import AMiner
from torch_geometric.nn import MetaPath2Vec

path = osp.join(osp.dirname(osp.realpath('__file__')), '../../data/AMiner')
dataset = AMiner(path)
data = dataset[0]
print(data)

reverse_edge_types = [(x[2], "{}_{}".format('rev', x[1]), x[0]) for x in data.edge_types]

transform = T.RandomLinkSplit(
    num_val=0.1,  # fraction of data held out for validation
    num_test=0.1, # fraction of data held out for test
    neg_sampling_ratio=0.0,
    edge_types = data.edge_types,
    rev_edge_types= reverse_edge_types
)
train_data, val_data, test_data = transform(data)

metapath = [
    ('author', 'writes', 'paper'),
    ('paper', 'published_in', 'venue'),
    ('venue', 'publishes', 'paper'),
    ('paper', 'written_by', 'author'),
]
device = 'cuda' if torch.cuda.is_available() else 'cpu'
model = MetaPath2Vec(train_data.edge_index_dict, embedding_dim=64,
                     metapath=metapath, walk_length=5, context_size=2,
                     walks_per_node=5, num_negative_samples=5,
                     sparse=True).to(device)

loader = model.loader(batch_size=128, shuffle=True, num_workers=0)

optimizer = torch.optim.SparseAdam(list(model.parameters()), lr=0.01)


def train(epoch, log_steps=100, eval_steps=2000):
    model.train()

    total_loss = 0
    for i, (pos_rw, neg_rw) in enumerate(loader):
        optimizer.zero_grad()
        loss = model.loss(pos_rw.to(device), neg_rw.to(device))
        loss.backward()
        optimizer.step()

        total_loss += loss.item()
        if (i + 1) % log_steps == 0:
            print((f'Epoch: {epoch}, Step: {i + 1:05d}/{len(loader)}, '
                   f'Loss: {total_loss / log_steps:.4f}'))
            total_loss = 0

        if (i + 1) % eval_steps == 0:
            acc = test()
            print((f'Epoch: {epoch}, Step: {i + 1:05d}/{len(loader)}, '
                   f'Acc: {acc:.4f}'))


# @torch.no_grad()

for epoch in range(1, 6):
    train(epoch)
    # acc = test()
    # print(f'Epoch: {epoch}, Accuracy: {acc:.4f}')

```
<img width="1322" alt="Screen Shot 2024-02-06 at 9 06 12 PM" src="https://github.com/pyg-team/pytorch_geometric/assets/26111084/bde7ac0a-f186-4542-b9da-3f04ca62709f">

**Comment by rusty1s** [comment, marked as answer]
Mh, works for me on latest PyG master. Which PyG version are you on?

**Reply by xubingze** [reply]
I tried to update my PyG then it turned out working fine for me.
